membership-attribute-service/app/services/AttributesMaker.scala

```diff
@@ -1,30 +1,41 @@
 package services
 
-import com.gu.memsub.Product
+import com.gu.memsub.{Benefit, Product}
 import com.gu.memsub.subsv2.SubscriptionPlan.AnyPlan
 import com.gu.memsub.subsv2.{GetCurrentPlans, Subscription}
 import com.typesafe.scalalogging.LazyLogging
 import models.Attributes
 import org.joda.time.LocalDate
+import com.github.nscala_time.time.OrderingImplicits.LocalDateOrdering
 
 class AttributesMaker extends LazyLogging {
 
   def attributes(identityId: String, subs: List[Subscription[AnyPlan]], forDate: LocalDate): Option[Attributes] = {
 
-    val groupedSubs: Map[Option[Product], List[Subscription[AnyPlan]]] = subs.groupBy(subscription => GetCurrentPlans(subscription, forDate).toOption.map(_.head.product))
-    val membershipSub = groupedSubs.getOrElse(Some(Product.Membership), Nil)
-    val contributionSub = groupedSubs.getOrElse(Some(Product.Contribution), Nil)
+    def getCurrentPlans(subscription: Subscription[AnyPlan]) = GetCurrentPlans(subscription, forDate).toOption
+    def getTopProduct(subscription: Subscription[AnyPlan]) = getCurrentPlans(subscription).map(_.head.product)
+    def getTopBenefitId(subscription: Subscription[AnyPlan]) = getCurrentPlans(subscription).map(_.head.charges.benefits.head.id)
+    def getTopPlanName(subscription: Subscription[AnyPlan]) = getCurrentPlans(subscription).map(_.head.name)
+    def getAllBenefits(subscription: Subscription[AnyPlan]) = getCurrentPlans(subscription).toList.flatMap(_.map(_.charges.benefits.list).list.flatten).toSet
 
-    val tier = membershipSub.headOption.map(sub => GetCurrentPlans(sub, forDate).toOption.map(_.head.charges.benefits.head.id)).flatten
-    val recurringContributionPaymentPlan = contributionSub.headOption.map(sub => GetCurrentPlans(sub, forDate).toOption.map(_.head.name)).flatten
-    val membershipJoinDate = membershipSub.map(_.startDate).headOption
+    val groupedSubs: Map[Option[Product], List[Subscription[AnyPlan]]] = subs.groupBy(getTopProduct)
```

Comment:
what does a key `None` mean? looks like subs that had an error while trying to read them.  does that make sense?